Add support for setting secret annotations

Crazybus opened the pull request:
Originally I thought it would be nice to just grab whatever annotations
the secretClaim had and stick them onto the secret. The main problem
with this is that there are all kind of auto generated metadata being
added by Kubernetes and could cause some weirdness.

Closes: #11

Crazybus committed:
Add support for setting secret annotations

Originally I thought it would be nice to just grab whatever annotations
the secretClaim had and stick them onto the secret. The main problem
with this is that there are all kind of auto generated metadata being
added by Kubernetes and could cause some weirdness.

Closes: #11

flaper87 commented on pkg/vault/controller.go:
```diff
@@ -107,6 +107,17 @@ func (ctrl *controller) tryRenewLease(id string) (*vaultapi.Secret, error) {
 	return ctrl.vclient.Sys().Renew(id, 0)
 }
 
+func mergeAnnotations(userAnnotations map[string]string, baseAnnotations map[string]string) map[string]string {
```
This function is simple and gets the job done but, sadly, it introduces extra value copies in the code that could be avoided. The function receives `userAnnotations` and `baseAnnotations` by value. This is great in the case of the former because the function changes the value and returns it but it's not needed for the latter. One may argue that the compiler should be smart enough to optimize these things out but I'm all for idiomatic and explicit coding. :smiley: 

One option would be to pass `baseAnnotations` as a pointer but that would still require us to copy the values of it when setting the keys in the copied `userAnnotations` map.

Below is an (non-tested) suggestion of how we could also write this function. Besides not copying the values, the implementation below removes the duplication of how the `leaseDuration` and `leaseExpiration` values are calculated. Arguably, this function could be more explicit and just accept pointers to `secret.LeaseDuration`, `secret.LeaseID`, and `claim.Spec.Annotations` but that's one more parameter for no real gain.

```go
func buildSecretAnnotations(secret *vaultapi.Secret, claim *kube.SecretClaim) map[string]string {
       leaseDuration := time.Duration(secret.LeaseDuration) * time.Second
       leaseExpiration := time.Now().Add(leaseDuration).Unix()

       annotations := map[string]string{
            LeaseIDKey:         secret.LeaseID,
	    LeaseExpirationKey: strconv.FormatInt(leaseExpiration, 10),
            RenewableKey:       strconv.FormatBool(secret.Renewable),
	}

	if claim.Spec.Annotations != nil {
	    for k, v := range claim.Spec.Annotations {
		annotations[k] = v
	    }
	}
	return annotations
}
```

Crazybus replied:
Great suggestion. I just pushed a new version using your suggestion.

flaper87 reviewed (changes requested):
I suggested an alternative implementation for the `mergeAnnotations` function. Looking forward to hear your thoughts :thought_balloon:

Crazybus committed:
Refactor secret annotation building to use pointers

Addresses: https://github.com/roboll/kube-vault-controller/pull/12/files#r352294337

flaper87 commented on pkg/vault/controller.go:
```diff
@@ -107,19 +109,37 @@ func (ctrl *controller) tryRenewLease(id string) (*vaultapi.Secret, error) {
+	}
+
+	for k, v := range claim.Spec.Annotations {
+		if _, exists := annotations[k]; !exists {
```
Nice one! Preventing ppl from overriding the default annotations :+1:

flaper87 reviewed (approved):
Love it!
